File: Lab2/Arrays.java
package labs;

public class Arrays {
    public static void main(String[] args) {
        // Original Array
        int[] numbers = new int[5];
        // Display contents
        display(numbers);
        // Insert numbers into array
        insert(numbers, 10, 0);
        insert(numbers, 20, 1);
        insert(numbers, 30, 2);
        insert(numbers, 40, 3);
        insert(numbers, 50, 4);
        System.out.println("\n");
        // Display contents
        display(numbers);
        System.out.println("\n");
        // Delete element at position 2
        delete(numbers, 2);
        // Display contents
        display(numbers);
        // Search for 20 in numbers array
        search(numbers, 20);
        // Search for 30 in numbers array
        search(numbers, 30);
        insert(numbers, 30, 2);
        // Reverse array
        reverse(numbers);
        // Display array
        display(numbers);
    }

    /**
     * Method to display contents of the array
     * @param array Array to display contents of
     */
    private static void display(int[] array) {
        for(int item : array) {
            System.out.println(item);
        }
    }

    /**
     * Method to insert a value at a given position
     * @param array Array to perform insertion on
     * @param value Value to insert
     * @param pos Position to insert at
     */
    private static void insert(int[] array, int value, int pos) {
        // Insert value at pos by replacing index with value
        for(int i = 0; i < array.length; i++) {
            if(i == pos)
                array[i] = value;
        }
    }

    /**
     * Method to delete an element at a given position
     * @param array Array to perform deletion on
     * @param pos Location in array to delete element from
     */
    private static void delete(int[] array, int pos) {
        // Delete element at pos by replacing with zero
        for(int i = 0; i < array.length; i++) {
            if (i == pos)
                array[i] = 0;
        }
    }

    /**
     * Method to search array for given value
     * @param array Array to perform search on
     * @param value Number to search array for
     */
    private static void search(int[] array, int value) {
        // Searching through array linearly, not efficient but works
        // isFound is set to true if the item is present
        boolean isFound = false;
        for(int i = 0; i < array.length; i++) {
            if(array[i] == value) {
                System.out.println("Item " + value + " found at position " + i);
                isFound = true;
            }
        }
        if(!isFound)
            System.out.println("Item " + value + " not found in array.");
    }

    /**
     * Method to reverse array
     * @param array Array to reverse
     */
    private static void reverse(int[] array) {
        // Reversing array
        for (int i = 0; i < array.length / 2; i++) {
            // Set temp to current item in array
            int temp = array[i];
            // Set current item in array to the last element
            // array.length = 5, - 1 to set it to the last item at index 4, then -i which is the current position
            array[i] = array[array.length-1-i];
            // Set the last element that hasn't been reversed to the element from the beginning
            array[array.length-1-i] = temp;
        }
    }
}

File: Lab3 Array Abstract Data Types/arrays.c
#include <stdio.h>

#define MAX 5

void init(int *array);
void insert(int *, int pos, int num);
void replace(int *, int pos, int num);
void delete (int *, int pos);
void reverse(int *);
void display(int *);
void search(int *, int num);

int main()
{
    int array[MAX];
    // Initialise Array
    init(array);
    // Insert numbers into array
    insert(array, 1, 5);
    insert(array, 2, 10);
    insert(array, 3, 3);
    insert(array, 4, 20);
    insert(array, 5, 8);

    printf("Array Contents after Insertion: \n");
    // Invoke display function
    display(array);

    // Replace numbers in the array
    replace(array, 0, 10);
    replace(array, 1, 5);
    replace(array, 2, 6);
    replace(array, 3, 200);
    replace(array, 4, 2);

    printf("Array Contents after numbers have been replaced: \n");
    // Invoke display function
    display(array);

    delete (array, 5);
    delete (array, 4);

    printf("Array Contents after Deletion: \n");
    // Invoke display function
    display(array);

    insert(array, 4, 200);
    insert(array, 5, 2);

    printf("Array Contents after Insertion: \n");
    // Invoke display function
    display(array);

    reverse(array);

    printf("Array Contents after Reversal: \n");
    // Invoke display function
    display(array);

    // Searching for numbers within the array
    search(array, 2);
    search(array, 5);
    search(array, 8);

    return 0;
}

void display(int *array)
{
    int idx;
    for (idx = 0; idx < MAX; idx++)
    {
        printf("%d\t", idx);
    }
    printf("\n");

    for (idx = 0; idx < MAX; idx++)
    {
        printf("%d\t", array[idx]);
    }
    printf("\n");
}

void init(int *array)
{
    int idx;
    for (idx = 0; idx < MAX; idx++)
    {
        array[idx] = 0;
    }
}

void insert(int *array, int pos, int num)
{
    int idx;
    for (idx = MAX - 1; idx >= pos; idx--)
    {
        array[idx] = array[idx - 1];
    }
    array[idx] = num;
}

void replace(int *array, int pos, int num)
{
    int idx;
    for (idx = 0; idx < MAX; idx++)
    {
        if (idx == pos)
            array[idx] = num;
    }
}

void delete (int *array, int pos)
{
    int idx;
    for (idx = pos; idx < MAX; idx++)
    {
        array[idx - 1] = array[idx];
    }
    array[idx - 1] = 0;
}

void reverse(int *array)
{
    int idx;
    for (idx = 0; idx < MAX / 2; idx++)
    {
        int temp = array[idx];
        array[idx] = array[MAX - 1 - idx];
        array[MAX - 1 - idx] = temp;
    }
}

void search(int *array, int num)
{
    int idx;
    for (idx = 0; idx < MAX; idx++)
    {
        if (array[idx] == num)
        {
            printf("%d found at position %d\n", num, idx + 1);
            return;
        }
    }
    if (idx == MAX)
        printf("%d not found in array\n", num);
}